contents/df_end.md

```diff
@@ -0,0 +1,224 @@
+## Index and Summarize
+
+```{=comment}
+summaries
+```
+
+Let's go back to the example data defined above
+
+```jl
+sco("""
+without_caption_label( # hide
+grades_2020()
+) # hide
+""")
+```
+
+To get a the **vector** for `name` back, we can use
+
+```{=comment}
+These two functions cannot be replaced by inline code due to
+`df.name` being converted to the same filename as `df[!, :name]` in Books.jl.
+I need to fix it.
+```
+
+```jl
+@sco JDS.names_grades1()
+```
+
+or
+
+```jl
+@sco JDS.names_grades2()
+```
+
+For the **row**, say the second row, we can use
+
+```jl
+sco("""
+df[2, :]
+df = DataFrame(df[2, :]) # hide
+without_caption_label(df) # hide
+""")
+```
+
+or create a function to give us row `i`
+
+```jl
+@sc JDS.grade_2020(1)
+```
+
+```jl
+sco("""
+without_caption_label( # hide
+grade_2020(2)
+) # hide
+""")
+```
+
+Continuing on this, we can also get only `names` for the first 2 rows:
+
+```jl
+@sco JDS.grades_indexing()
+```
+
+If we assume that all names in the table are unique, we can also write a function to obtain the grade for a person via their `name`.
+To do this, we convert the table back to one of Julia's basic data structures which is capable of creating a mappings, namely dictionaries:
+
+```jl
+@sc grade_2020("")
+```
+
+```jl
+sco("""
+grade_2020("Bob")
+""")
+```
+
+which works because `zip` loops through `df.name` and `df.grade_2020` at the same time like a zipper:
+
+```jl
+sco("""
+df = grades_2020()
+collect(zip(df.name, df.grade_2020))
+""")
+```
+
+However, converting a DataFrame to a Dictionary is only useful when the elements are unique.
+When that is not the case, it is time to `filter` (@sec:filter).
+
+## Filter and Subset {#sec:filter}
+
+Continuing from the earlier mentioned table.
+
+```jl
+sco("""
+without_caption_label( # hide
+grades_2020()
+) # hide
+""")
+```
+
+We can filter rows by using `filter(f::Function, df)`.
+This function is very similar to the function `filter(f::Function, V::Vector)` from Julia itself.
+Working with a function `f` for filtering can be a bit difficult to use in practice, but it is very powerful.
+As a simple example, we can create a function which checks whether it's input equals "Alice":
+
+```jl
+@sc JDS.equals_alice("")
+```
+
+```jl
+sco("""
+equals_alice("Bob")
+""")
+```
+
+```jl
+sco("""
+equals_alice("Alice")
+""")
+```
+
+With this function, we can now filter out all the rows for which `name` equals "Alice"
+
+```jl
+sco("""
+without_caption_label( # hide
+filter(:name => equals_alice, grades_2020())
+) # hide
+""")
+```
+
+Note that this doesn't only work for DataFrames, but also for vectors:
```

Comment:
See you are now using the `:`, we really need to standardize a common practice here.